File: src/utils/sessionFile.ts
import { invoke } from '@tauri-apps/api/core'
import { isTauri, mockInvoke } from '../mock-tauri'
import type { Session } from './sessionSchema'

/**
 * The Session file's command boundary. The Rust side owns `session.json`
 * (atomic write, ~500 ms debounce, a last flush on quit) and merges the
 * window frame in; the renderer only reads the file back at launch and hands
 * over its part of the state whenever it changes. Outside Tauri the Folder
 * fixture answers both commands.
 */

/** The raw file contents, or null when there is no Session to restore. */
export function readSessionFile(): Promise<unknown> {
  return isTauri() ? invoke<unknown>('read_session') : mockInvoke<unknown>('read_session')
}

export function updateSessionFile(session: Session): Promise<void> {
  const args = { session }
  return isTauri() ? invoke<void>('update_session', args) : mockInvoke<void>('update_session', args)
}

File: src-tauri/src/commands/session_file.rs
use crate::session::{self, SessionState};
use serde_json::Value;
use tauri::Manager;

/// The Session file as it is on disk, or null when there is none to restore.
/// The renderer validates the schema (an unknown `version` is ignored and
/// rewritten on its side).
#[tauri::command]
pub fn read_session(app_handle: tauri::AppHandle) -> Option<Value> {
    session::read_session_file(app_handle.state::<SessionState>().path())
}

/// The renderer's part of the Session changed; the file follows after the debounce.
#[tauri::command]
pub fn update_session(app_handle: tauri::AppHandle, session: Value) -> Result<(), String> {
    if !session.is_object() {
        return Err("Session must be an object".to_owned());
    }
    session::update_renderer_session(&app_handle, session);
    Ok(())
}
